drivers/i2c/i2c_renesas_ra_iic.c

```diff
@@ -0,0 +1,552 @@
+					      &data->iic_master_ext_cfg.clock_settings);
+		data->iic_master_ext_cfg.timeout_mode = IIC_MASTER_TIMEOUT_MODE_SHORT;
+		data->iic_master_ext_cfg.timeout_scl_low = IIC_MASTER_TIMEOUT_SCL_LOW_ENABLED;
+
+		data->fsp_config.p_extend = &data->iic_master_ext_cfg;
+		break;
+	default:
+		LOG_ERR("%s: Invalid I2C speed rate: %d", __func__, data->fsp_config.rate);
+		return -ENOTSUP;
+	}
+
+	fsp_err = R_IIC_MASTER_Open(&data->ctrl, &data->fsp_config);
+	__ASSERT(fsp_err == FSP_SUCCESS, "%s: Open iic master failed. FSP_ERR=%d", __func__,
+		 fsp_err);
+
+	config->irq_config_func(dev);
+
+	return 0;
+}
+
+static void calc_iic_master_bitrate(const struct i2c_ra_iic_config *config, uint32_t total_brl_brh,
+				    uint32_t brh, uint32_t divider,
+				    struct ra_iic_master_bitrate *result)
+{
+	const uint32_t noise_filter_stage = config->noise_filter_stage;
+	const double rise_time_ns = config->rise_time_ns / RA_IIC_MASTER_DIV_TIME_NS;
+	const double fall_time_ns = config->fall_time_ns / RA_IIC_MASTER_DIV_TIME_NS;
+	const uint32_t requested_duty = config->duty_cycle_percent;
+	const uint32_t peripheral_clock = R_FSP_SystemClockHzGet(FSP_PRIV_CLOCK_PCLKB);
+	uint32_t constant_add = 0;
+
+	/* A constant is added to BRL and BRH in all formulas. This constand is 3 + nf when CKS ==
+	 * 0, or 2 + nf when CKS != 0.
+	 */
+	if (divider == 0) {
+		constant_add = 3 + noise_filter_stage;
+	} else {
+		/* All dividers other than 0 use an addition of 2 + noise_filter_stages. */
+		constant_add = 2 + noise_filter_stage;
+	}
+
+	/* Converts all divided numbers to double to avoid data loss. */
+	uint32_t divided_pclk = (peripheral_clock >> divider);
+
+	result->bitrate = 1 / ((total_brl_brh + 2 * constant_add) / divided_pclk + rise_time_ns +
+			       fall_time_ns);
+	result->duty = 100 * ((rise_time_ns + ((brh + constant_add) / divided_pclk)) /
+			      (rise_time_ns + fall_time_ns +
+			       ((total_brl_brh + 2 * constant_add)) / divided_pclk));
+	result->divider = divider;
+	result->brh = brh;
+	result->brl = total_brl_brh - brh;
+	result->duty_error_percent =
+		(result->duty > requested_duty ? result->duty - requested_duty
+					       : requested_duty - result->duty) /
+		requested_duty;
+
+	LOG_DBG("%s: [input] total_brl_brh[%d] brh[%d] divider[%d]"
+		" [output] bitrate[%u] duty[%u] divider[%u] brh[%u] brl[%u] duty_error_percent[%u]\n",
+		__func__, total_brl_brh, brh, divider, result->bitrate, result->duty,
+		result->divider, result->brh, result->brl, result->duty_error_percent);
+}
+
+static void calc_iic_master_clock_setting(const struct device *dev, const uint32_t fsp_i2c_rate,
+					  iic_master_clock_settings_t *clk_cfg)
+{
+	const struct i2c_ra_iic_config *config = dev->config;
+	const uint32_t noise_filter_stage = config->noise_filter_stage;
+	const double rise_time_ns = config->rise_time_ns / RA_IIC_MASTER_DIV_TIME_NS;
+	const double fall_time_ns = config->fall_time_ns / RA_IIC_MASTER_DIV_TIME_NS;
```

Comment:
Could we move those divisions in `i2c_ra_iic_config_##index` definition so it will happen in compile time?